drivers/hmis/app/models/hmis/workflow_execution/engine.rb

```diff
@@ -115,6 +121,17 @@ def visit_node(node, user)
       case node
       when Hmis::WorkflowDefinition::Task
         step = instance.steps.find_or_initialize_by(node: node)
+
+        # If the step has already been completed, it may be re-openable
+        if step.status == 'completed'
+          step_side_effects = step.node.trigger_config&.map { |config| config['message'] } || []
+
+          # but _only_ if it didn't have any irreversible side effects
+          if (Hmis::Ce::ReferralMessageHandler::IRREVERSIBLE & step_side_effects).any? # rubocop:disable Style/IfUnlessModifier
```

Comment:
This is breaking the boundary that the `drivers/hmis/app/models/hmis/workflow_execution` models shouldn't have any CE-specific code.

Is there a way we can encapsulate this further, for example something like `@message_handler.contains_irreversible_triggers?(step)`. That feels a little wonky still, but better than referencing a Ce module from within the engine.

Reply:
Hm yeah that's a very good point. Thanks for the catch!

I'm proposing `@message_handler.contains_irreversible_messages?(messages)`, which still seems a little messy, but it keeps knowledge of "step" shape out of the message handler and CE logic out of the execution engine. Let me know if it looks okay to you.

Reply:
looks good to me, thanks!